mpv-player/mpv
# [RFC] Add stream for DVB using libdvbv5

As a base I've used dvbv5-zap (from the v4l-utils project) and tried to
maintain the same design for an MPV stream.

This dvbv5 stream should obsolete the dvbin stream but there are some
breaking changes.
* the channels.conf file format has changes (dvbin uses VDR format,
  dvdv5 uses DVBV5).
* dvbin allows you to pass the adapter/card number in the url like:
  dvb://1@Some Channel. I don't see the added value of this.
* dvbin automatically select a frontend that should be used. I'm not
  sure if that's always a good thing.
* requires an extra external library (libdvbv5), seems to be available
  for most linux distributions.

There a some great advantages of libdvb5.
* maintained by the v4l-utils/linux-media community.
* a whole lot more configuration options.
* includes a channels.conf parser (dvbin has one of it's own)
* less code to maintain in mpv.

As I envision to keep using the v4l-utils dvbv5-zap sources to fix
bugs/add new features in the future I've done the following to allow
easy comparison.
* don't prefix n_vpid, n_apid with cfg_.
* keep the functions in the same order as dvbv5-zap source code
* use the same structure for timeouts (dvbv5-zap uses alarm/signal, I've
  ported this to a thread)

Some things weren't possible or feasible.
* converted tabs to spaces
* replaced goto err with return
* a lot of chatty printfs

The following is still todo
* The remote dvbv5 server code doesn't seem to do anything. Might be
  incompatible with mpv in the current state.
* Test for memory leaks
* Would be really nice use the EPG (EIT) to add program information
  to the meta data.
* I've only tested DVB-T2 with a cheap USB tuner. There are some
  specific DVB-S features which I couldn't test.
* Remove/deprecate dvbin, use dvb:// url instead for dvbv5.
* I've found some minor issues with dvbv5-zap (fixed in MPV) which
  I'm planning to PR as well.
* Check if the allowed ranges for the parameters make sense.
* Document how to generate channels.conf.dvbv5 and where it should be
  placed.

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.
